src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpResponseParser.cs

```diff
@@ -33,27 +28,37 @@ internal static class WinHttpResponseParser
             var response = new HttpResponseMessage();
             bool stripEncodingHeaders = false;
 
+            // Create a single buffer to use for all subsequent WinHttpQueryHeaders string interop calls.
+            // This buffer is the length needed for WINHTTP_QUERY_RAW_HEADERS_CRLF, which includes the status line
+            // and all headers separated by CRLF, so it should be large enough for any individual status line or header queries.
+            int bufferLength = GetResponseHeaderCharBufferLength(requestHandle, Interop.WinHttp.WINHTTP_QUERY_RAW_HEADERS_CRLF);
+            char[] buffer = new char[bufferLength];
+
             // Get HTTP version, status code, reason phrase from the response headers.
-            string version = GetResponseHeaderStringInfo(requestHandle, Interop.WinHttp.WINHTTP_QUERY_VERSION);
-            if (string.Compare("HTTP/1.1", version, StringComparison.OrdinalIgnoreCase) == 0)
-            {
-                response.Version = HttpVersion.Version11;
-            }
-            else if (string.Compare("HTTP/1.0", version, StringComparison.OrdinalIgnoreCase) == 0)
-            {
-                response.Version = HttpVersion.Version10;
-            }
-            else
-            {
-                response.Version = HttpVersion.Unknown;
-            }
+
+            int versionLength = GetResponseHeader(requestHandle, Interop.WinHttp.WINHTTP_QUERY_VERSION, buffer);
+            response.Version =
+                CharArrayHelpers.EqualsOrdinalAsciiIgnoreCase("HTTP/1.1", buffer, 0, versionLength) ? HttpVersion.Version11 :
+                CharArrayHelpers.EqualsOrdinalAsciiIgnoreCase("HTTP/1.0", buffer, 0, versionLength) ? HttpVersion.Version10 :
+                HttpVersion.Unknown;
 
             response.StatusCode = (HttpStatusCode)GetResponseHeaderNumberInfo(
                 requestHandle,
                 Interop.WinHttp.WINHTTP_QUERY_STATUS_CODE);
-            response.ReasonPhrase = GetResponseHeaderStringInfo(
-                requestHandle,
-                Interop.WinHttp.WINHTTP_QUERY_STATUS_TEXT);
+
+            int reasonPhraseLength = GetResponseHeader(requestHandle, Interop.WinHttp.WINHTTP_QUERY_STATUS_TEXT, buffer);
+            if (reasonPhraseLength > 0)
+            {
+                string reasonPhrase;
+
+                // If it's a known reason phrase, use the known reason phrase instead of allocating a new string.
+                if (!TryGetKnownReasonPhrase(response.StatusCode, buffer, 0, reasonPhraseLength, out reasonPhrase))
+                {
+                    reasonPhrase = new string(buffer, 0, reasonPhraseLength);
+                }
+
+                response.ReasonPhrase = reasonPhrase; // It's OK if this is null.
```

Comment:
I removed the comment (it was a holdover from an earlier iteration that I should have been deleted).